src/program.ts

```diff
@@ -0,0 +1,194 @@
+    }
+  });
+}
+
+function loadProjectId(program: Command, cmd: Command) {
+  const opts = cmd.optsWithGlobals();
+
+  if (opts.apiKey?.startsWith(API_KEY_PAK_PREFIX)) {
+    // Parse the key and ensure we can access the specified Project ID
+    const projectId = RestClient.projectIdFromKey(opts.apiKey);
+    program.setOptionValue('projectId', projectId);
+
+    if (opts.projectId !== -1 && opts.projectId !== projectId) {
+      error(
+        'The specified API key cannot be used to perform operations on the specified project.'
+      );
+      info(
+        `The API key you specified is tied to project #${projectId}, you tried to perform operations on project #${opts.projectId}.`
+      );
+      info(
+        'Learn more about how API keys in Tolgee work here: https://tolgee.io/platform/account_settings/api_keys_and_pat_tokens'
+      );
+      process.exit(1);
+    }
+  }
+}
+
+function validateOptions(cmd: Command) {
+  const opts = cmd.optsWithGlobals();
+  if (opts.projectId === -1) {
+    error(
+      'No Project ID have been specified. You must either provide one via --project-id, or by setting up a `.tolgeerc` file.'
+    );
+    info(
+      'Learn more about configuring the CLI here: https://tolgee.io/tolgee-cli/project-configuration'
+    );
+    process.exit(1);
+  }
+
+  if (!opts.apiKey) {
+    error(
+      'No API key has been provided. You must either provide one via --api-key, or login via `tolgee login`.'
+    );
+    process.exit(1);
+  }
+}
+
+async function preHandler(prog: Command, cmd: Command) {
+  const NO_KEY_COMMANDS = ['login', 'logout', 'extract'];
+
+  if (!NO_KEY_COMMANDS.includes(topLevelName(cmd))) {
+    await loadApiKey(prog, cmd);
+    loadProjectId(prog, cmd);
+    validateOptions(cmd);
+
+    const opts = cmd.optsWithGlobals();
+    const client = new RestClient({
+      apiUrl: opts.apiUrl,
+      apiKey: opts.apiKey,
+      projectId: opts.projectId,
+    });
+
+    cmd.setOptionValue('client', client);
+  }
+
+  // Apply verbosity
+  setDebug(prog.opts().verbose);
+}
+
+export function loadEnvironmentalVariables(program: Command) {
+  const options: BaseOptions = program.optsWithGlobals();
+  const envFilePath = path.resolve(process.cwd(), options.env);
+
+  if (fs.existsSync(envFilePath)) {
+    dotenv.config({ path: envFilePath });
+
+    /** Sets the option value if it was not specified by the user. */
+    const setOptionValue = (key: string, value: unknown) => {
+      if (program.getOptionValueSourceWithGlobals(key) !== 'cli') {
+        program.setOptionValue(key, value);
+      }
+    };
+
+    if (process.env.TOLGEE_API_KEY) {
+      setOptionValue('apiKey', process.env.TOLGEE_API_KEY);
+    }
+    if (process.env.TOLGEE_API_URL) {
+      setOptionValue('apiUrl', parseUrlArgument(process.env.TOLGEE_API_URL));
+    }
+    if (process.env.TOLGEE_PROJECT_ID) {
+      setOptionValue(
+        'projectId',
+        parseProjectId(process.env.TOLGEE_PROJECT_ID)
+      );
+    }
```

Comment:
I'm unsure how to feel about that. I'm aware we can't fetch env variables be fetched before because we need to get `--env`, however doing manually the env checks does duplicate the places where it needs to be handled and is prone to desync...

Iterating over the list of available options stored in `program` is probably better

Reply:
I can do a followup refactoring PR to handle loading the configuration in one place.

The reason I did not iterate here, because different options need different parsing function, like the `TOLGEE_PROJECT_ID` needs the `parseProjectId()`.

Reply:
Isn't the parse function already stored and accessible via the list of available options?

Reply:
Would you be able to show me a code snippet on how we could improve this function? I might just be slow today, but it would help me for sure.

Reply:
It's possible to get the list of options that can be specified on a given command using the [`options` field](https://github.com/tj/commander.js/blob/v11.1.0/typings/index.d.ts#L297); considering this I wonder if it's not possible to use this (along with `parent` since commands are structured in a tree-like structure) to get all the options that can be specified, and check if the option can use an env variable, if so use its parser and its `name()` to populate the command options - see [this](https://github.com/tj/commander.js/blob/v11.1.0/typings/index.d.ts#L109-L110)

Reply:
I haven't find a way to iterate over the options and set the env variables that way. It is not possible to get whether an option can use an env variable.

Reply:
Hello, sorry for the late reply, I unfortunately got sick 🤧

I was able to iterate over all the options using this snippet:
```js
function processCommand (command: Command) {
  for (const option of command.options) {
    console.log(option.attributeName(), option.envVar)
  }

  if (command.parent) {
    processCommand(command.parent)
  }
}
```